index.bs

```diff
@@ -4629,14 +4887,33 @@ partial interface MLGraphBuilder {
         - *SumSquare*: Compute the sum of the square of all the input values along the axes.
 </div>
 
+<details open algorithm>
+  <summary>
+    To <dfn for="MLGraphBuilder">calculate reduction output sizes</dfn>, given a [=/list=] of unsigned integers |inputShape|, a optional [=/list=] of unsigned integers |axes|, and [=/boolean=] |keepDimensions|, perform the following steps. They return a new [=/list=] of unsigned integers.
+  </summary>
+  <div class=algorithm-steps>
+    1. Let |inputSize| be |inputShape|'s [=list/size=].
+    1. If |axes| is not given, let |axes| be [=the range=] 0 to |inputSize|, exclusive.
+    1. If |keepDimensions| is true, then:
+        1. Let |outputShape| be a [=list/clone=] of |inputShape|.
+        1. [=list/For each=] |axis| of |axes|:
+            1. Set |outputShape|[|axis|] to 1.
+    1. Otherwise:
+        1. Let |outputShape| be an empty [=/list=].
+        1. [=list/For each=] |index| in [=the range=] 0 to |inputSize|, exclusive:
+            1. If |axes| does not [=list/contain=] |index|, then [=list/append=] |inputShape|[|index|].
```

Comment:
Yep, this looks right, handling the empty `axes` too (nop case), the all axes case (scalar result case), and degenerate (but not illegal) cases like axes in arbitrary order [4,2,3] and repeats [2,3,2].